compiler/test/dotty/tools/dotc/ParallelTesting.scala

```diff
@@ -0,0 +1,1076 @@
+              val errorMsg = target.buildInstructions(reporter.errorCount, reporter.warningCount)
+              addFailureInstruction(errorMsg)
+              failTarget(target)
+              reporter.echo(errorMsg)
+              reporter.flushToFile()
+            }
+          }
+
+          case target @ SeparateCompilationTarget(dir, flags, outDir) => {
+            val compilationUnits = target.compilationUnits
+            val reporters = compilationUnits.map(files => compile(files.filter(isCompilable), flags, false, outDir))
+            val errorCount = reporters.foldLeft(0) { (acc, reporter) =>
+              if (reporter.errorCount > 0) {
+                val errorMsg = target.buildInstructions(reporter.errorCount, reporter.warningCount)
+                addFailureInstruction(errorMsg)
+                reporter.echo(errorMsg)
+                reporter.flushToFile()
+              }
+
+              acc + reporter.errorCount
+            }
+
+            completeCompilation(errorCount)
+
+            if (errorCount > 0) failTarget(target)
+          }
+        }
+
+      }
+    }
+  }
+
+  private final class RunTest(targets: List[Target], times: Int, threadLimit: Option[Int], suppressAllOutput: Boolean)
+  extends Test(targets, times, threadLimit, suppressAllOutput) {
+    private def verifyOutput(checkFile: JFile, dir: JFile, target: Target, warnings: Int) = try {
+        // Do classloading magic and running here:
+        import java.net.{ URL, URLClassLoader }
+        import java.io.ByteArrayOutputStream
+        val ucl = new URLClassLoader(Array(dir.toURI.toURL))
+        val cls = ucl.loadClass("Test")
+        val meth = cls.getMethod("main", classOf[Array[String]])
+
+        val printStream = new ByteArrayOutputStream
+        Console.withOut(printStream) {
+          meth.invoke(null, Array("jvm")) // partest passes at least "jvm" as an arg
+        }
+
+        val outputLines = printStream.toString("utf-8").lines.toArray
+        val checkLines = Source.fromFile(checkFile).getLines.toArray
+
+        def linesMatch =
+          outputLines
+          .zip(checkLines)
+          .forall { case (x, y) => x == y }
+
+        if (outputLines.length != checkLines.length || !linesMatch) {
+          // Print diff to files and summary:
+          val diff = outputLines.zip(checkLines).map { case (act, exp) =>
+            DiffUtil.mkColoredCodeDiff(exp, act, true)
+          }.mkString("\n")
+          val msg = s"\nOutput from run test '$checkFile' did not match expected, output:\n$diff\n"
+          echo(msg)
+          addFailureInstruction(msg)
+
+          // Print build instructions to file and summary:
+          val buildInstr = target.buildInstructions(0, warnings)
+          addFailureInstruction(buildInstr)
+
+          // Fail target:
+          failTarget(target)
+        }
+    }
+    catch {
+      case _: NoSuchMethodException =>
+        echo(s"\ntest in '$dir' did not contain a main method")
```

Comment:
Well then `test in '$dir' did not contain a main method` is still the truth right?

Not sure how you're proposing it to be checked otherwise.